src/DocumentStore.zig

```diff
@@ -588,6 +600,32 @@ fn getOrLoadBuildFile(self: *DocumentStore, uri: Uri) ?*BuildFile {
     return gop.value_ptr.*;
 }
 
+pub fn doubleCheckPotentialBuildFileUri(self: *DocumentStore, handle: *Handle) error{OutOfMemory}!?Uri {
```

Comment:
If we have a handle with multiple potential build files but none of them are associated to the handle, then this functions returns `null`. Every subsequent call will then do its entire work again and then still return `null`. There should be some kind of caching. This should also help with reducing contention on the handle mutex.

Reply:
I'd say the issue is when there are none, otherwise if they all've been ran the first one sticks.

I'm thinking maybe passing an optional handle uri to getOrLoadBuildFile to do the association after the file has been ran (for every build file other than the first one). It'd be easier to pass a *Handle, but having an uri and calling getHandle on it ensures the handle is still valid.

Reply:
Isn't it possible to return null when searching through multiple potential build files because of the `if (!ok) handle.associated_build_file = null;` line? If `!ok` then we return `null` and any subsequent call would also return `null`.

Reply:
Equipped with the knowledge of how `for() {} else` works I think correcting the else to `else true;` should handle this condition;